src/YieldDisburser.sol

```diff
@@ -167,127 +134,130 @@ contract YieldDisburser is OwnableUpgradeable {
      */
     function getVotingPowerForPeriod(uint256 _start, uint256 _end, address _account) external view returns (uint256) {
-        // Checking if the start time is before the end time, if the end time is after the current
-        // block and if the user has ever held $BREAD comparing the first mint to the end of the interval
-        if (_start > _end) revert StartMustBeBeforeEnd();
+        if (_start >= _end) revert StartMustBeBeforeEnd();
+        if (_end > block.number) revert EndAfterCurrentBlock();
 
-        if (_end > Time.blockNumber()) revert EndAfterCurrentBlock();
+        // Initialized as the checkpoint count, but later used to track checkpoint index
+        uint32 _currentCheckpointIndex = BREAD.numCheckpoints(_account);
+        if (_currentCheckpointIndex == 0) revert NoCheckpointsForAccount();
 
-        uint32 latestCheckpointPos = BREAD.numCheckpoints(_account);
-        if (latestCheckpointPos == 0) revert NoCheckpointsForAccount();
+        // No voting power if the first checkpoint is after the end of the interval
+        Checkpoints.Checkpoint208 memory _currentCheckpoint = BREAD.checkpoints(_account, 0);
+        if (_currentCheckpoint._key > _end) return 0;
 
-        if (BREAD.checkpoints(_account, 0)._key > _end) return 0;
+        // Find the latest checkpoint that is within the interval
+        do {
+            --_currentCheckpointIndex;
+            _currentCheckpoint = BREAD.checkpoints(_account, _currentCheckpointIndex);
+        } while (_currentCheckpoint._key > _end);
 
-        uint256 value;
-        uint48 lastKey = type(uint48).max;
-        Checkpoints.Checkpoint208 memory checkpoint;
+        // Initialize voting power with the latest checkpoint thats within the interval (or nearest to it)
+        uint48 _latestKey = _currentCheckpoint._key < _start ? uint48(_start) : _currentCheckpoint._key;
+        uint256 _totalVotingPower = _currentCheckpoint._value * (_end - _latestKey);
 
-        // Find the latest checkpoint that is within the interval
-        while (lastKey > _end) {
-            latestCheckpointPos--;
-            checkpoint = BREAD.checkpoints(_account, latestCheckpointPos);
-            lastKey = checkpoint._key;
-        }
+        for (uint32 i = _currentCheckpointIndex; i > 0;) {
+            // Latest checkpoint voting power is calculated when initializing `_totalVotingPower`, so we pre-decrement the index here
+            _currentCheckpoint = BREAD.checkpoints(_account, --i);
 
-        // We are now at a position where the checkpoint is within the interval or the one nearest it
-        value = checkpoint._value;
-        lastKey = uint48((lastKey < _start ? _start : lastKey));
-        uint256 votingPowerTotal = value * (_end - lastKey);
-        // If there's a single checkpoint in the interval, return the voting power from the interval
-        if (latestCheckpointPos == 0 || lastKey < _start) return votingPowerTotal;
```

Comment:
its redundant because of the ternary response above `lastKey < _start ? _start : lastKey` which means `lastKey` will always be at least `== _start`

Reply:
Then I got something wrong but essentially if the first checkpoint t is before the start we don't need to continue the function logic but return the voting power, so it should have been if lastKey == _start , return votingPowerTotal 

Does that make sense?

Reply:
yea, but unless im missing something, this loop should still return the correct value in that case:

```
_totalVotingPower += _currentCheckpoint._value * (_latestKey - _currentCheckpoint._key);
````
it adds the value here

```
            // At the start of the interval, deduct voting power accrued before the interval and return the total
            if (_currentCheckpoint._key <= _start) {
                _totalVotingPower -= _currentCheckpoint._value * (_start - _currentCheckpoint._key);
                break;
            }
```
and subtracts the same value here

since it's an edge case I think this is fine?

Reply:
I believe this is actually the average case, as most users will not be minting bread every month, but minting once and then voting for many cycles. 

We should not be fetching another checkpoint and adding it at all,  and the subtraction there is for the diff between the start and the second closest checkpoint to the interval, not the closest checkpoint to the interval. The diff between the closest checkpoint in the interval to the start is never subtracted.

Reply:
This condition is only for the off chance the latest checkpoint in the interval is at the same block as the start of the interval, in which case we don't need to do any calculation outside the initialization of '_votingPower'. From what I can tell, the results are the same with the refactored function. Maybe we're not talking about the same thing?

Reply:
it's if the latest checkpoint is before the beginning of the interval, we basically adjust the lastKey so that it considers this checkpoint as the relevant checkpoint for the whole interval.

Reply:
```mermaid
sequenceDiagram
User ->> Bread : Mint 10 bread at block 20 
User ->> Bread : Mint 10 bread at block 30
User ->> YieldDistributer : Vote at block 50  
YieldDistributer ->>  YieldDistributer : Distribute for period block 40 -60
YieldDistributer ->> YieldDistributer : Consider User voting power as 20 bread  * 20 blocks 
```

Reply:
Ok so looking over at this , the loop basically corrects this , but we don't have to go into the loop in the first place. Because in this cast , _latestKey == _start , we will subtract what we have added inside the loop exactly always. It's just a bit awkward I guess? but we can keep this

Reply:
resolved with cbb4cd4